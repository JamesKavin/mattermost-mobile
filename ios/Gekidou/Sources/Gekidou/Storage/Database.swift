//
//  Database.swift
//  Gekidou
//
//  Created by Miguel Alatzar on 8/20/21.
//

import Foundation
import SQLite3
import SQLite

// TODO: This should be exposed to Objective-C in order to handle
// any Database throwable methods.
enum DatabaseError: Error {
    case OpenFailure(_ dbPath: String)
    case MultipleServers
    case NoResults(_ query: String)
    case NoDatabase(_ serverUrl: String)
    case InsertError(_ statement: String)
}

extension DatabaseError: LocalizedError {
    var errorDescription: String? {
        switch self {
        case .OpenFailure(dbPath: let dbPath):
            return "Error opening database: \(dbPath)"
        case .MultipleServers:
            return "Cannot determine server URL as there are multiple servers"
        case .NoResults(query: let query):
            return "No results for query: \(query)"
        case .NoDatabase(serverUrl: let serverUrl):
            return "No database for server: \(serverUrl)"
        case .InsertError(statement: let statement):
            return "Insert error: \(statement)"
        }
    }
}

public class Database: NSObject {
    internal let DEFAULT_DB_NAME = "app.db"
    internal var DEFAULT_DB_PATH: String
    internal var defaultDB: OpaquePointer? = nil
    
    internal var serversTable = Table("Servers")
    internal var systemTable = Table("System")
    internal var teamTable = Table("Team")
    internal var myTeamTable = Table("MyTeam")
    internal var channelTable = Table("Channel")
    internal var channelInfoTable = Table("ChannelInfo")
    internal var channelMembershipTable = Table("ChannelMembership")
    internal var myChannelTable = Table("MyChannel")
    internal var myChannelSettingsTable = Table("MyChannelSettings")
    internal var postTable = Table("Post")
    internal var postsInChannelTable = Table("PostsInChannel")
    internal var postsInThreadTable = Table("PostsInThread")
    internal var postMetadataTable = Table("PostMetadata")
    internal var reactionTable = Table("Reaction")
    internal var fileTable = Table("File")
    internal var emojiTable = Table("CustomEmoji")
    
    @objc public static let `default` = Database()
    
    override private init() {
        let appGroupId = Bundle.main.object(forInfoDictionaryKey: "AppGroupIdentifier") as! String
        let sharedDirectory = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupId)!
        let databaseUrl = sharedDirectory.appendingPathComponent("databases/\(DEFAULT_DB_NAME)")
        
        DEFAULT_DB_PATH = databaseUrl.path
    }
    
    @objc public func getOnlyServerUrlObjc() -> String {
        do {
            return try getOnlyServerUrl()
        } catch {
            print(error)
            return ""
        }
    }
    
    public func getOnlyServerUrl() throws -> String {
        let db = try Connection(DEFAULT_DB_PATH)
        let url = Expression<String>("url")
        let query = serversTable.select(url)
        
        var serverUrl: String?
        for result in try db.prepare(query) {
            if (serverUrl != nil) {
                throw DatabaseError.MultipleServers
            }
            
            serverUrl = try result.get(url)
        }
        
        if (serverUrl != nil) {
            return serverUrl!
        }
    
        throw DatabaseError.NoResults(query.asSQL())
    }
    
    internal func getDatabaseForServer(_ serverUrl: String) throws -> Connection {
        let db = try Connection(DEFAULT_DB_PATH)
        let url = Expression<String>("url")
        let dbPath = Expression<String>("db_path")
        let query = serversTable.select(dbPath).where(url == serverUrl)
        
        if let result = try db.pluck(query) {
            let path = try result.get(dbPath)
            return try Connection(path)
        }
        
        throw DatabaseError.NoResults(query.asSQL())
    }
    
    internal func queryCurrentUserId(_ serverUrl: String) throws -> String {
        let db = try getDatabaseForServer(serverUrl)
        
        let idCol = Expression<String>("id")
        let valueCol = Expression<String>("value")
        let query = systemTable.where(idCol == "currentUserId")
        
        if let result = try db.pluck(query) {
            return try result.get(valueCol).replacingOccurrences(of: "\"", with: "")
        }
        
        throw DatabaseError.NoResults(query.asSQL())
    }
    
    private func json(from object:Any?) -> String? {
        guard let object = object, let data = try? JSONSerialization.data(withJSONObject: object, options: []) else {
            return nil
        }
        return String(data: data, encoding: String.Encoding.utf8)
    }
}
