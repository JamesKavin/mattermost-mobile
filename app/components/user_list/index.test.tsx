// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Database from '@nozbe/watermelondb/Database';
import React from 'react';

import {renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';

import UserList from '.';

describe('components/channel_list_row', () => {
    let database: Database;
    const user: UserProfile = {
        id: '1',
        create_at: 1111,
        update_at: 1111,
        delete_at: 0,
        username: 'johndoe',
        auth_service: '',
        email: 'john@doe.com',
        nickname: '',
        first_name: '',
        last_name: '',
        position: '',
        roles: '',
        locale: '',
        notify_props: {
            channel: 'true',
            comments: 'never',
            desktop: 'mention',
            desktop_sound: 'true',
            email: 'true',
            first_name: 'true',
            mention_keys: '',
            push: 'mention',
            push_status: 'away',
        },
    };
    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
    });

    it('should show no results', () => {
        const wrapper = renderWithEverything(
            <UserList
                profiles={[user]}
                testID='UserListRow'
                currentUserId={'1'}
                teammateNameDisplay={'johndoe'}
                handleSelectProfile={() => {
                    // noop
                }}
                fetchMore={() => {
                    // noop
                }}
                loading={true}
                selectedIds={{}}
                showNoResults={true}
                tutorialWatched={true}
            />,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should show results no tutorial', () => {
        const wrapper = renderWithEverything(
            <UserList
                profiles={[user]}
                testID='UserListRow'
                currentUserId={'1'}
                teammateNameDisplay={'johndoe'}
                handleSelectProfile={() => {
                    // noop
                }}
                fetchMore={() => {
                    // noop
                }}
                loading={true}
                selectedIds={{}}
                showNoResults={true}
                tutorialWatched={true}
            />,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should show results and tutorial', () => {
        const wrapper = renderWithEverything(
            <UserList
                profiles={[user]}
                testID='UserListRow'
                currentUserId={'1'}
                teammateNameDisplay={'johndoe'}
                handleSelectProfile={() => {
                    // noop
                }}
                fetchMore={() => {
                    // noop
                }}
                loading={true}
                selectedIds={{}}
                showNoResults={false}
                tutorialWatched={false}
            />,
            {database},
        );

        expect(wrapper.toJSON()).toMatchSnapshot();
    });
});
