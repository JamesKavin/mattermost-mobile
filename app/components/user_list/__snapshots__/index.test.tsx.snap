// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/channel_list_row should show no results 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
  data={
    [
      {
        "data": [
          {
            "auth_service": "",
            "create_at": 1111,
            "delete_at": 0,
            "email": "john@doe.com",
            "first_name": "",
            "id": "1",
            "last_name": "",
            "locale": "",
            "nickname": "",
            "notify_props": {
              "channel": "true",
              "comments": "never",
              "desktop": "mention",
              "desktop_sound": "true",
              "email": "true",
              "first_name": "true",
              "mention_keys": "",
              "push": "mention",
              "push_status": "away",
            },
            "position": "",
            "roles": "",
            "update_at": 1111,
            "username": "johndoe",
          },
        ],
        "first": true,
        "id": "J",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  initialNumToRender={15}
  keyExtractor={[Function]}
  keyboardDismissMode="on-drag"
  keyboardShouldPersistTaps="always"
  maxToRenderPerBatch={16}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={60}
  stickyHeaderIndices={[]}
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  testID="UserListRow.section_list"
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "rgba(63,67,80,0.08)",
              "height": 24,
              "justifyContent": "center",
              "paddingLeft": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 12,
                "fontWeight": "600",
                "lineHeight": 16,
              }
            }
          >
            J
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
                "height": 58,
                "overflow": "hidden",
                "paddingHorizontal": 20,
              }
            }
            testID="create_direct_message.user_list.user_item.1"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "color": "#3f4350",
                    "flexDirection": "row",
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "borderRadius": 21.5,
                    "height": 42,
                    "width": 42,
                  }
                }
                testID="create_direct_message.user_list.user_item.1.profile_picture"
              >
                <Icon
                  name="account-outline"
                  size={24}
                  style={
                    {
                      "color": "rgba(63,67,80,0.48)",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                [
                  {
                    "flex": 1,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "paddingHorizontal": 10,
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    {
                      "color": "#3f4350",
                      "fontFamily": "OpenSans",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "height": 24,
                      "lineHeight": 24,
                      "maxWidth": "80%",
                    }
                  }
                  testID="create_direct_message.user_list.user_item.1.display_name"
                >
                  johndoe
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Icon
                color="rgba(63,67,80,0.32)"
                name="circle-outline"
                size={28}
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <ActivityIndicator
          color="#1c58d9"
          size="large"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`components/channel_list_row should show results and tutorial 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
  data={
    [
      {
        "data": [
          {
            "auth_service": "",
            "create_at": 1111,
            "delete_at": 0,
            "email": "john@doe.com",
            "first_name": "",
            "id": "1",
            "last_name": "",
            "locale": "",
            "nickname": "",
            "notify_props": {
              "channel": "true",
              "comments": "never",
              "desktop": "mention",
              "desktop_sound": "true",
              "email": "true",
              "first_name": "true",
              "mention_keys": "",
              "push": "mention",
              "push_status": "away",
            },
            "position": "",
            "roles": "",
            "update_at": 1111,
            "username": "johndoe",
          },
        ],
        "first": true,
        "id": "J",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  initialNumToRender={15}
  keyExtractor={[Function]}
  keyboardDismissMode="on-drag"
  keyboardShouldPersistTaps="always"
  maxToRenderPerBatch={16}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={60}
  stickyHeaderIndices={[]}
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  testID="UserListRow.section_list"
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "rgba(63,67,80,0.08)",
              "height": 24,
              "justifyContent": "center",
              "paddingLeft": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 12,
                "fontWeight": "600",
                "lineHeight": 16,
              }
            }
          >
            J
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
                "height": 58,
                "overflow": "hidden",
                "paddingHorizontal": 20,
              }
            }
            testID="create_direct_message.user_list.user_item.1"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "color": "#3f4350",
                    "flexDirection": "row",
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "borderRadius": 21.5,
                    "height": 42,
                    "width": 42,
                  }
                }
                testID="create_direct_message.user_list.user_item.1.profile_picture"
              >
                <Icon
                  name="account-outline"
                  size={24}
                  style={
                    {
                      "color": "rgba(63,67,80,0.48)",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                [
                  {
                    "flex": 1,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "paddingHorizontal": 10,
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    {
                      "color": "#3f4350",
                      "fontFamily": "OpenSans",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "height": 24,
                      "lineHeight": 24,
                      "maxWidth": "80%",
                    }
                  }
                  testID="create_direct_message.user_list.user_item.1.display_name"
                >
                  johndoe
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Icon
                color="rgba(63,67,80,0.32)"
                name="circle-outline"
                size={28}
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <ActivityIndicator
          color="#1c58d9"
          size="large"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`components/channel_list_row should show results no tutorial 1`] = `
<RCTScrollView
  ListEmptyComponent={[Function]}
  ListFooterComponent={[Function]}
  contentContainerStyle={
    {
      "flexGrow": 1,
    }
  }
  data={
    [
      {
        "data": [
          {
            "auth_service": "",
            "create_at": 1111,
            "delete_at": 0,
            "email": "john@doe.com",
            "first_name": "",
            "id": "1",
            "last_name": "",
            "locale": "",
            "nickname": "",
            "notify_props": {
              "channel": "true",
              "comments": "never",
              "desktop": "mention",
              "desktop_sound": "true",
              "email": "true",
              "first_name": "true",
              "mention_keys": "",
              "push": "mention",
              "push_status": "away",
            },
            "position": "",
            "roles": "",
            "update_at": 1111,
            "username": "johndoe",
          },
        ],
        "first": true,
        "id": "J",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  initialNumToRender={15}
  keyExtractor={[Function]}
  keyboardDismissMode="on-drag"
  keyboardShouldPersistTaps="always"
  maxToRenderPerBatch={16}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={true}
  renderItem={[Function]}
  scrollEventThrottle={60}
  stickyHeaderIndices={[]}
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
  testID="UserListRow.section_list"
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "backgroundColor": "#ffffff",
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "rgba(63,67,80,0.08)",
              "height": 24,
              "justifyContent": "center",
              "paddingLeft": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#3f4350",
                "fontFamily": "OpenSans-SemiBold",
                "fontSize": 12,
                "fontWeight": "600",
                "lineHeight": 16,
              }
            }
          >
            J
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <View
            style={
              {
                "flex": 1,
                "flexDirection": "row",
                "height": 58,
                "overflow": "hidden",
                "paddingHorizontal": 20,
              }
            }
            testID="create_direct_message.user_list.user_item.1"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "color": "#3f4350",
                    "flexDirection": "row",
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "borderRadius": 21.5,
                    "height": 42,
                    "width": 42,
                  }
                }
                testID="create_direct_message.user_list.user_item.1.profile_picture"
              >
                <Icon
                  name="account-outline"
                  size={24}
                  style={
                    {
                      "color": "rgba(63,67,80,0.48)",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                [
                  {
                    "flex": 1,
                    "flexDirection": "column",
                    "justifyContent": "center",
                    "paddingHorizontal": 10,
                  },
                  {
                    "opacity": 1,
                  },
                ]
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                  }
                }
              >
                <Text
                  ellipsizeMode="tail"
                  numberOfLines={1}
                  style={
                    {
                      "color": "#3f4350",
                      "fontFamily": "OpenSans",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "height": 24,
                      "lineHeight": 24,
                      "maxWidth": "80%",
                    }
                  }
                  testID="create_direct_message.user_list.user_item.1.display_name"
                >
                  johndoe
                </Text>
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "justifyContent": "center",
                }
              }
            >
              <Icon
                color="rgba(63,67,80,0.32)"
                name="circle-outline"
                size={28}
              />
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
    >
      <View
        style={
          {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <ActivityIndicator
          color="#1c58d9"
          size="large"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
